llvm/lib/CodeGen/PreISelIntrinsicLowering.cpp

```diff
@@ -66,6 +71,27 @@ struct PreISelIntrinsicLowering {
 
 } // namespace
 
+template <class T>
+static bool forEachCall(Function &Intrin, T Callback) {
+  // Lowering all intrinsics in a function will delete multiple uses, so we
+  // can't use an early-inc-range. In case some remain, we don't want to look
+  // at them again. Unfortunately, Value::UseList is private, so we can't use a
+  // simple Use**. If LastUse is null, we the next use to consider is
+  // Intrin.use_begin(), otherwise it's LastUse->getNext().
+  Use *LastUse = nullptr;
+  bool Changed = false;
+  while (!Intrin.use_empty() && (!LastUse || LastUse->getNext())) {
+    Use *U = LastUse ? LastUse->getNext() : &*Intrin.use_begin();
+    bool Removed = false;
+    auto CI = dyn_cast<CallInst>(U->getUser());
+    if (CI && CI->getCalledOperand() == &Intrin)
```

Comment:
These conditions should always be true. The IR verifier checks that.

Reply:
There's the case of assume-like intrinsics, which are permitted ([example](https://godbolt.org/z/z46o9h9jv) with debug record). I think that's also the reason this check is found elsewhere in the file.

Reply:
Huh, I didn't know that. I guess you could still drop the `getCalledOperand` check.

Reply:
Changed. I couldn't come up with a case where this invariant does not hold. It might make sense to simplify the other checks later.

Reply:
But metadata uses are not part of the use list.